sway-lib-std/src/ufp64.sw

```diff
@@ -0,0 +1,247 @@
+
+        let s_mul_o = s_value * o_value;
+        let res_u128 = s_mul_o >> 32;
+        if res_u128.upper != 0 {
+            // panic on overflow
+            revert(0);
+        }
+
+        UFP64 {
+            value: res_u128.lower
+        }
+    }
+}
+
+impl core::ops::Divide for UFP64 {
+    /// Divide a UFP64 by a UFP64. Panics if divisor is zero.
+    pub fn divide(self, divisor: Self) -> Self {
+    let zero = ~UFP64::zero();
+        assert(divisor != zero);
+
+        let denominator = ~U128::from(0, ~Self::denominator());
+        let self_u128 = ~U128::from(0, self.value);
+        let divisor_u128 = ~U128::from(0, divisor.value);
+
+        let res_u128 = self_u128 * denominator / divisor_u128;
+        if res_u128.upper != 0 {
+            // panic on overflow
+            revert(0);
+        }
+        UFP64 {
+            value: res_u128.lower
+        }
+    }
+}
+
+
+
+impl UFP64 {
+
+    pub fn from_uint(uint: u64) -> UFP64 {
+        UFP64 {
+            value: ~Self::denominator() * uint,
+        }
+    }
+}
+
+impl UFP64 {
+
+    pub fn recip(number: UFP64) -> Self {
+        let one = ~UFP64::from_uint(1);
+        
+        let res = one / number;
+        res
+    }
+}
+
+impl UFP64 {
+
+    pub fn trunc(self) -> Self {
+        UFP64 {
+            value: (self.value >> 32) << 32
+        }
+    }
+}
+
+impl UFP64 {
+
+    pub fn floor(self) -> Self {
+        return self.trunc();
+    }
+
+    pub fn fract(self) -> Self {
+        UFP64 {
+            value: (self.value << 32) >> 32
+        }
+    }
+}
+
+impl UFP64 {
+
+    pub fn ceil(self) -> Self {
+        if self.fract().value != 0 {
+            let res = self.trunc() + ~UFP64::from_uint(1);
+            return res;
+        }
+        return self;
+    }
+}
+
+impl UFP64 {
+
+    pub fn round(self) -> Self {
+        let floor = self.floor();
+        let ceil = self.ceil();
+        let diff_self_floor = self - floor;
+        let diff_ceil_self = ceil - self;
+        if diff_self_floor < diff_ceil_self {
+            return floor;
+        } else {
+            return ceil;
+        }
+    }  
+}
+
+impl Root for UFP64 {
+    fn sqrt(self) -> Self {
+        let nominator_root = self.value.sqrt();
+        let nominator = nominator_root << 16;
+        UFP64 {
+            value: nominator
+        }
+    }
+}
+
+impl Exponentiate for UFP64 {
+    pub fn pow(self, exponent: Self) -> Self {
+        let denominator_power = 32;
+        let exponent_int = exponent.value >> denominator_power;
+        let nominator_pow = ~U128::from(0, self.value).pow(~U128::from(0, exponent_int));
+        let nominator = nominator_pow >> denominator_power*(exponent_int - 1);
+        if nominator.upper != 0 {
+            // panic on overflow
+            revert(0);
+        }
+        UFP64 {
+            value: nominator.lower
+        }
+    }
+}
+
+// TODO: uncomment and change accordingly, when signed integers will be added
```

Comment:
I'm unsure about things like this. I generally do not want things to just linger around because of tech debt however if this is going to be implemented soon then it's okay.

Reply:
Deleted this for now